libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -354,10 +357,16 @@ void VoteManager::cleanupVotes(uint64_t pbft_round) {
     UniqueLock lock(verified_votes_access_);
     auto it = verified_votes_.begin();
     while (it != verified_votes_.end() && it->first < pbft_round) {
-      for (auto const& step : it->second) {
-        for (auto const& voted_value : step.second) {
-          for (auto const& v : voted_value.second.second) {
+      for (const auto& step : it->second) {
+        for (const auto& voted_value : step.second) {
+          for (const auto& v : voted_value.second.second) {
             db_->removeVerifiedVoteToBatch(v.first, batch);
+            LOG(log_dg_) << "Remove verified vote " << v.first << " for round " << v.second->getRound()
+                         << ". PBFT round " << pbft_round;
+            if (v.second->getType() == cert_vote_type) {
+              // The verified vote may reward vote
+              AddRewardVote(v.second);
```

Comment:
This seems little bit complicated. Why dont you simply set `reward_votes_` with cert_votes of pbft block right after we process new pbft block ? Also at that point you can  delete saved rewards_votes for pbft block that we just processed. So that would be one place where we set reward_votes + we would add new reward vote when received through network

Reply:
reward_votes_ is for missing cert votes in previous period, and in previous period means the PBFT block has been finalized. So we have at least 2t+1 cert votes for the pervious period finalized PBFT block already. If we receive more cert votes for previous period, we could add it to previous period sync_block immediately(you mentioned that's not good for performance because that cause many write in DB), or we add it to reward_votes_ and later add to sync_block in bunch.

When propose PBFT block, we do add reward_votes_(new cert votes of previous period) and delete saved rewards_votes.

That cannot be processed reward_votes only when propose PBFT block, here is an example:
1. node1 finalize PBFT block B1 with cert votes {v1, v2, v3}, then node1 receive v4 for B1 after finalize B1, and put v4 to rewards_votes
2. When node1 propose B2-1, node1 add rewards_votes v4 for B1, and B2 reward votes are {v1, v2, v3, v4}
3. In the meanwhile, node2 propose B2-2 with reward votes {v1, v2, v3, v4, v5}
4. Node2 proposed B2-2 win the leader of blocks. At the time, node1 doesn't have v5, node1 will receive v5 after it propose B1-1. So still need add v5 later

The code for a special case:
1. node finalize B1 with cert votes {v1, v2, v3} in certify state
2. Then node goes to next voting state, and receive v4 for B1 from peers
3. When node goes to next PBFT round, that will delete all votes for previous rounds. After we introduce reward votes, cannot delete previous cert votes like v4.

Reply:
`When propose PBFT block, we do add reward_votes_(new cert votes of previous period) and delete saved rewards_votes.` - why would you do that ? When you propose PBFT block for period 10, you should not add or delete rewards_votes_. When you propose PBFT block, you should simply get all votes from your locally saved rewards_votes_ and put it as block rewards votes. Thats it - nothing else. Then, after some PBFT block for period 10 is chosen(can be yours or someone else's) and it is processed(finalized) you should take it's cert votes and put it into the rewards_votes_

`That cannot be processed reward_votes only when propose PBFT block, here is an example:` - I dont understand what are you trying to say here ? I mean, the sentence does not make sense - arent you missing some words ?

Also the example you provided is not very clear to me

Reply:
Here is an example why need add the verified votes to reward votes queue:
1. In certify stage, node receive enough cert votes and finalize PBFT block B1 with cert votes {v1, v2, v3}
2. Then node goes to next voting stage in the same round, and node receive additional cert vote v4 for B1. node will add v4 in votes queue here https://github.com/Taraxa-project/taraxa-node/blob/reward_cert_votes/libraries/core_libs/network/src/tarcap/packets_handlers/vote_packet_handler.cpp#L55
3. node receives enough next votes and goes to next PBFT round. node will clean up all votes (include v4) < current PBFT round here https://github.com/Taraxa-project/taraxa-node/blob/reward_cert_votes/libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp#L275
We cannot delete v4 here, need add v4 in reward votes